src/solvers/refinement/string_refinement.cpp

```diff
@@ -336,6 +336,32 @@ static bool has_char_pointer_subtype(const typet &type, const namespacet &ns)
   return false;
 }
 
+/// \param type: a type
+/// \param ns: name space
+/// \return true if a subtype is of type string_typet
+static bool has_string_subtype(const typet &type, const namespacet &ns)
+{
+  if(type == string_typet())
+    return true;
+
+  if(type.id() == ID_struct || type.id() == ID_union)
+  {
+    const struct_union_typet &struct_type = to_struct_union_type(type);
+    for(const auto &comp : struct_type.components())
+    {
+      if(has_string_subtype(comp.type(), ns))
+        return true;
+    }
+  }
+
+  for(const auto &t : type.subtypes())
```

Comment:
This doesn't mean subtype in the Java sense, but rather the arguments of a type in the algebraic datatype sense -- for example, the "subtypes" of a struct are the types of its components (therefore the struct special case above is redundant), the "subtype" of a pointer is the type it points to... and so on.

Reply:
@smowton 
> therefore the struct special case above is redundant

I'm not sure it is: in a struct the subtypes appear in the components, which is a `named_sub` for the `irept` and not in `sub` which is what `subtypes` returns

Reply:
I added to the documentation the precision that we mean subtype in the algebraic sense

Reply:
Ah so it is. No problem then.